examples/src/main/java/com/mapbox/navigation/examples/core/MapMatchingActivity.kt

```diff
@@ -0,0 +1,346 @@
+        mapInstanceState = savedInstanceState?.getParcelable(MAP_INSTANCE_STATE_KEY)
+        directionsRoute = getRouteFromBundle(savedInstanceState)
+    }
+
+    fun startLocationUpdates() {
+        val request = LocationEngineRequest.Builder(INTERVAL_MILLISECONDS)
+            .setFastestInterval(FASTEST_INTERVAL_MILLISECONDS)
+            .setPriority(LocationEngineRequest.PRIORITY_HIGH_ACCURACY)
+            .build()
+
+        localLocationEngine.requestLocationUpdates(
+            request,
+            locationEngineCallback,
+            Looper.getMainLooper()
+        )
+
+        if (directionsRoute == null) {
+            localLocationEngine.getLastLocation(locationEngineCallback)
+        }
+    }
+
+    private fun stopLocationUpdates() {
+        localLocationEngine.removeLocationUpdates(locationEngineCallback)
+    }
+
+    private val routesReqCallback = object : RoutesRequestCallback {
+        override fun onRoutesReady(routes: List<DirectionsRoute>) {
+            if (routes.isNotEmpty()) {
+                val coordinates = mutableListOf<Point>()
+                routes[0].legs()?.forEach { leg ->
+                    leg.steps()?.forEach { step ->
+                        step.maneuver().location().run {
+                            coordinates.add(Point.fromLngLat(longitude(), latitude()))
+                        }
+                    }
+                }
+
+                locationComponent?.lastKnownLocation?.let {
+                    val origin = Point.fromLngLat(it.longitude, it.latitude)
+                    coordinates.add(0, origin)
+                    coordinates.add(destination!!)
+
+                    fetchRoute(coordinates)
+                }
+            }
+        }
+
+        override fun onRoutesRequestFailure(throwable: Throwable, routeOptions: RouteOptions) {
+            Timber.e("DirectionsAPI route request failure %s", throwable.toString())
+        }
+
+        override fun onRoutesRequestCanceled(routeOptions: RouteOptions) {
+            Timber.d("DirectionsAPI route request canceled")
+        }
+    }
+
+    private val tripSessionStateObserver = object : TripSessionStateObserver {
+        override fun onSessionStateChanged(tripSessionState: TripSessionState) {
+            when (tripSessionState) {
+                STARTED -> stopLocationUpdates()
+                STOPPED -> {
+                    startLocationUpdates()
+                    navigationMapboxMap?.removeRoute()
+                    updateCameraOnNavigationStateChange(false)
+                }
+            }
+        }
+    }
+
+    private val offRouteObserver = object : OffRouteObserver {
+        override fun onOffRouteStateChanged(offRoute: Boolean) {
+            if (offRoute) {
+                ifNonNull(
+                    locationComponent?.lastKnownLocation,
+                    destination
+                ) { location, destination ->
+                    fetchRouteWithDirectionsApi(location.toPoint(), destination)
```

Comment:
When there's an `offRoute` internally we `reroute` by default using the Directions API https://github.com/mapbox/mapbox-navigation-android/blob/67552b9a95bc4ac7f1f5ba570a22702dbcb5d73b/libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt#L616-L630 which fires off a route request. In this example, that route request is canceled because of https://github.com/mapbox/mapbox-navigation-android/blob/4f5e3e5e0411e5d430abe0ab73f1b445c571c2f8/examples/src/main/java/com/mapbox/navigation/examples/core/MapMatchingActivity.kt#L129 That also sets the primary route as above-mentioned. That's not optimal though as we're trying to do an unnecessary request. In order to avoid that, a custom `RerouteController` should be used instead `OffRouteObserver` - that's exactly what `createInternalOffRouteObserver` / `reroute` do so no need for this. Note that adding `setRoutes` as part of `Callback<MapMatchingResponse>` would work but it's not the best approach as would mean an unnecessary extra request, it's more brittle because if the underlying canceling logic changes it may not work anymore (order matters) and it's not intuitive IMO from a developer standpoint as it forces you to know how the API works behind the scenes while using `RerouteController` you're in total control of re-route handling. That'd also help us confirming the need of exposing `MapboxRouteOptionsUpdater` publicly https://github.com/mapbox/mapbox-navigation-android/pull/3429#issue-466870266 and showcasing how to use it if needed.